### PR TITLE
Make ProfiledTensorType hashable

### PR DESCRIPTION
**Pull request by Krovatkin**

### CODE REVIEW

File: torch/csrc/jit/argument_spec.cpp [hunk 1]
```diff
@@ -284,5 +284,29 @@ void ArgumentSpecCreator::specializeTypes(
   }
 }
 
+template <typename F, typename T>
```
Comment (zdevito) [inline torch/csrc/jit/argument_spec.cpp]: I don't get this, it seems like it is just being used to convert a ScalarType to an optional int8_t. Why not write the function that does directly that? The templates are hard to follow.

**Review by zdevito — APPROVED** [review]
This looks fine, apart from a nit, but where is this going to be used?

**Comment by Krovatkin** [comment]
@zdevito this will hopefully replace using `CompleteArgumentSpec`s as keys to cache glow and tvm kernels: https://github.com/pytorch/tvm/blob/master/torch_tvm/compiler.cpp#L227

File: torch/csrc/jit/argument_spec.cpp (hunk 1 above)
Comment (Krovatkin) [reply torch/csrc/jit/argument_spec.cpp]: Initially, I thought I'd have more enum classes that I'd need to convert, but then I forgot to clean it up. My bad.
